In [45]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

Đọc file

In [46]:
df = pd.read_csv('archive/DataScientist.csv',index_col='index')
df.head(5)

Unnamed: 0                          Job Title  \
index                                                  
0               0              Senior Data Scientist   
1               1  Data Scientist, Product Analytics   
2               2               Data Science Manager   
3               3                       Data Analyst   
4               4             Director, Data Science   

                    Salary Estimate  \
index                                 
0      $111K-$181K (Glassdoor est.)   
1      $111K-$181K (Glassdoor est.)   
2      $111K-$181K (Glassdoor est.)   
3      $111K-$181K (Glassdoor est.)   
4      $111K-$181K (Glassdoor est.)   

                                         Job Description  Rating  \
index                                                              
0      ABOUT HOPPER\n\nAt Hopper, we’re on a mission ...     3.5   
1      At Noom, we use scientifically proven methods ...     4.5   
2      Decode_M\n\nhttps://www.decode-m.com/\n\nData ...    -1.0   
3      Sapphire Digital seeks a dynamic and driven mi...     3.4   
4      Director, Data Science - (200537)\nDescription...     3.4   

                          Company Name       Location      Headquarters  \
index                                                                     
0                          Hopper\n3.5   New York, NY  Montreal, Canada   
1                         Noom US\n4.5   New York, NY      New York, NY   
2                             Decode_M   New York, NY      New York, NY   
3                Sapphire Digital\n3.4  Lyndhurst, NJ     Lyndhurst, NJ   
4      United Entertainment Group\n3.4   New York, NY      New York, NY   

                         Size  Founded  Type of ownership  \
index                                                       
0       501 to 1000 employees     2007  Company - Private   
1      1001 to 5000 employees     2008  Company - Private   
2           1 to 50 employees       -1            Unknown   
3        201 to 500 employees     2019  Company - Private   
4         51 to 200 employees     2007  Company - Private   

                        Industry                  Sector  \
index                                                      
0                Travel Agencies        Travel & Tourism   
1      Health, Beauty, & Fitness       Consumer Services   
2                             -1                      -1   
3                       Internet  Information Technology   
4        Advertising & Marketing       Business Services   

                        Revenue               Competitors Easy Apply  
index                                                                 
0      Unknown / Non-Applicable                        -1         -1  
1      Unknown / Non-Applicable                        -1         -1  
2      Unknown / Non-Applicable                        -1       True  
3      Unknown / Non-Applicable      Zocdoc, Healthgrades         -1  
4      Unknown / Non-Applicable  BBDO, Grey Group, Droga5         -1

In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3909 entries, 0 to 4379
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         3909 non-null   int64  
 1   Job Title          3909 non-null   object 
 2   Salary Estimate    3909 non-null   object 
 3   Job Description    3909 non-null   object 
 4   Rating             3909 non-null   float64
 5   Company Name       3909 non-null   object 
 6   Location           3909 non-null   object 
 7   Headquarters       3909 non-null   object 
 8   Size               3909 non-null   object 
 9   Founded            3909 non-null   int64  
 10  Type of ownership  3909 non-null   object 
 11  Industry           3909 non-null   object 
 12  Sector             3909 non-null   object 
 13  Revenue            3909 non-null   object 
 14  Competitors        3909 non-null   object 
 15  Easy Apply         3909 non-null   object 
dtypes: float64(1), int64(2),

# Xử lí dữ liệu

#  Cleaning

Thay thế các giá trị -1 bằng NaN

In [48]:
df.replace(['-1'], [np.nan], inplace=True)
df.replace(['-1.0'], [np.nan], inplace=True)
df.replace([-1], [np.nan], inplace=True)

Kiểm tra xem còn giá trị khác không?

In [49]:
df.isnull().sum()

Unnamed: 0              0
Job Title               0
Salary Estimate         0
Job Description         0
Rating                409
Company Name            0
Location                0
Headquarters          240
Size                  229
Founded               977
Type of ownership     229
Industry              546
Sector                546
Revenue               229
Competitors          2760
Easy Apply           3745
dtype: int64

# Tách cột Salary Estimate thành 2 cột mới: minimum salary và maximum salary

In [50]:
df_salary = df['Salary Estimate'].str.split("-",expand=True,)

minimum_salary = df_salary[0]
minimum_salary = minimum_salary.str.replace('K',' ')


maximum_salary = df_salary[1].str.replace('(Glassdoor est.)', ' ')
maximum_salary = maximum_salary.str.replace('(', ' ')
maximum_salary = maximum_salary.str.replace(')', ' ')
maximum_salary = maximum_salary.str.replace('K', ' ')
maximum_salary = maximum_salary.str.replace('Employer est.', ' ')
maximum_salary = maximum_salary.str.replace('Per Hour', ' ')

maximum_salary = maximum_salary.str.replace('$', ' ').fillna(0).astype(int)
minimum_salary = minimum_salary.str.replace('$', ' ').fillna(0).astype(int)

df['Minimum Salary'] = minimum_salary
df['Maximum Salary'] = maximum_salary

df.drop('Salary Estimate',axis = 1,inplace = True)

# Ta cũng tách cột Revenue thành 2 cột mới: Maximum Revenue và Minimum Revenue

Đầu tiên tiến hành định dạng lại dữ liệu để có thể xử lí

In [51]:
# Thay thế các giá trị unknow thành none
df['Revenue'] = df['Revenue'].replace('Unknown / Non-Applicable', None)

In [52]:
df['Revenue'] = df['Revenue'].str.replace('$', ' ')
df['Revenue'] = df['Revenue'].str.replace('(USD)', ' ')
df['Revenue'] = df['Revenue'].str.replace('(', ' ')
df['Revenue'] = df['Revenue'].str.replace(')', ' ')
df['Revenue'] = df['Revenue'].str.replace(' ', '')

In [53]:
df['Revenue'] = df['Revenue'].str.replace('2to5billion', '2billionto5billion')
df['Revenue'] = df['Revenue'].str.replace('5to10billion ', '5billionto10billion ')
df['Revenue'] = df['Revenue'].replace('million', ' ')
df['Revenue'] = df['Revenue'].replace('10+billion', '10billionto11billion')
df['Revenue'] = df['Revenue'].str.replace('Lessthan1million', '0millionto1million')

In [54]:
df['Revenue'] = df['Revenue'].str.replace('million', ' ')
df['Revenue'] = df['Revenue'].str.replace('billion', '000 ')

Tiến hành chia dữ liệu thành 2 cột

In [55]:
Revenue = df['Revenue'].str.split("to",expand=True)
df['Minimum Revenue'] = Revenue[0]
df['Maximum Revenue'] = Revenue[1]

# Đưa dữ liệu về định dạng int64
df['Maximum Revenue'] = pd.to_numeric(df['Maximum Revenue'])
df['Minimum Revenue'] = pd.to_numeric(df['Minimum Revenue'])

#Xóa cột Revenue cũ 
df.drop('Revenue',1,inplace=True)


In [56]:
# Xóa phần dữ liệu thừa phía sau tên công ty
df['Company Name'] = df['Company Name'].str.replace('\n.*', ' ')

# Tách cột size thành cột Minimum Size và Maximum Size

In [57]:
df['Size'].value_counts()

10000+ employees           993
51 to 200 employees        563
1001 to 5000 employees     553
1 to 50 employees          550
201 to 500 employees       418
501 to 1000 employees      307
5001 to 10000 employees    219
Unknown                     77
Name: Size, dtype: int64

In [58]:
# Xóa employees
df['Size'] = df['Size'].str.replace('employees', '')

In [59]:
df['Size'] = df['Size'].str.replace('+', 'plus')
df['Size'] = df['Size'].replace('Unknown', None)

df['Size'] = df['Size'].str.replace('10000plus', '10000 to 10001')


In [60]:
size = df['Size'].str.split("to",expand=True)
df['Minimum Size'] = size[0]
df['Maximum Size'] = size[1]
df.drop('Size',1,inplace = True)

df['Minimum Size'] = pd.to_numeric(df['Minimum Size'])
df['Maximum Size'] = pd.to_numeric(df['Maximum Size'])


In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3909 entries, 0 to 4379
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         3909 non-null   int64  
 1   Job Title          3909 non-null   object 
 2   Job Description    3909 non-null   object 
 3   Rating             3500 non-null   float64
 4   Company Name       3909 non-null   object 
 5   Location           3909 non-null   object 
 6   Headquarters       3669 non-null   object 
 7   Founded            2932 non-null   float64
 8   Type of ownership  3680 non-null   object 
 9   Industry           3363 non-null   object 
 10  Sector             3363 non-null   object 
 11  Competitors        1149 non-null   object 
 12  Easy Apply         164 non-null    object 
 13  Minimum Salary     3909 non-null   int32  
 14  Maximum Salary     3909 non-null   int32  
 15  Minimum Revenue    2517 non-null   float64
 16  Maximum Revenue    2517 

# 1. Đánh giá các công ty lớn

Tiêu chí đánh giá: rating, Revenue, Size, Salary Estimate

+ Rating >= 3

+ Minimum size > 10000

+ Minimum Revenue > 10000

+ Salary min > 100

=> Biết được các công ty có quy mô lớn và đang phát triển mạnh. Qua đó tìm kiếm cơ hội làm ăn và phát triển các mối quan hệ đối với các công ty này.

Ở đây không sử dụng cột founded (tức năm thành lập) vì nhiều công ty dù được thành lập muộn hơn nhưng biết nắm bắt thời cơ và có phương hướng phát triển tốt nên đã trở thành các công ty mạnh mẽ.

In [62]:
df_big_company = df[(df['Rating'] >= 3.0) & (df['Minimum Revenue'] >= 10000.0) & (df['Minimum Revenue'] == 10000.0) & (df['Minimum Salary'] >=100)]
df_big_company

Unnamed: 0                                          Job Title  \
index                                                                  
20             20            Data Scientist, Disney+ Personalization   
27             27                                Lead Data Scientist   
29             29                           Principal Data Scientist   
53             51                               CIB - Data Scientist   
56             54                      Lead Data Scientist - 66201BR   
...           ...                                                ...   
4181         3711  Cyber Security Research Scientist 3 - Sr. Appl...   
4221         3751  Data Scientist/Programmer, Sr. Analyst - Abbot...   
4222         3752       Data Scientist Sr. Associate - SALES SCIENCE   
4228         3758   Data Scientist Associate Sr (DADS06) CCB ML & AI   
4231         3761  Quantum Data Scientist (Industrial-Process sec...   

                                         Job Description  Rating  \
index                                                              
20     Job Summary:Comprised of Disney's internationa...     4.0   
27     Description: Its a new day in health care.\n\n...     3.3   
29     Description: Its a new day in health care.\n\n...     3.3   
53     *CIB Applied AI & Machine Learning - Digital &...     3.9   
56     Description: Its a new day in health care.\n\n...     3.3   
...                                                  ...     ...   
4181   Job Description\n\nImportant Note: During the ...     3.6   
4221   Abbott is a global healthcare leader that help...     3.6   
4222   The Chase Sales Science team supports the exec...     3.9   
4228   *Data Scientist Sr. Associate (AI/ML)****Chase...     3.9   
4231   Introduction\nAs a Data Scientist at IBM, you ...     3.7   

                Company Name      Location       Headquarters  Founded  \
index                                                                    
20          Walt Disney Co.   New York, NY        Burbank, CA   1923.0   
27                    Aetna   New York, NY       Hartford, CT   1853.0   
29                    Aetna   New York, NY       Hartford, CT   1853.0   
53           JPMorgan Chase   New York, NY       New York, NY   1799.0   
56                    Aetna   New York, NY       Hartford, CT   1853.0   
...                      ...           ...                ...      ...   
4181            Wells Fargo     Irving, TX  San Francisco, CA   1852.0   
4221    Abbott Laboratories   Columbus, OH    Abbott Park, IL   1888.0   
4222   JPMorgan Chase & Co.   Columbus, OH       New York, NY   1799.0   
4228         JPMorgan Chase   Columbus, OH       New York, NY   1799.0   
4231                    IBM   Columbus, OH         Armonk, NY   1911.0   

      Type of ownership                                  Industry  \
index                                                               
20     Company - Public  Motion Picture Production & Distribution   
27     Company - Public                        Insurance Carriers   
29     Company - Public                        Insurance Carriers   
53     Company - Public     Investment Banking & Asset Management   
56     Company - Public                        Insurance Carriers   
...                 ...                                       ...   
4181   Company - Public                     Banks & Credit Unions   
4221   Company - Public          Health Care Services & Hospitals   
4222   Company - Public     Investment Banking & Asset Management   
4228   Company - Public     Investment Banking & Asset Management   
4231   Company - Public                               IT Services   

                       Sector                           Competitors  \
index                                                                 
20                      Media  News Corp, WarnerMedia, NBCUniversal   
27                  Insurance                                   NaN   
29                  Insurance           

=> Có thể thấy data ban đầu gồm 3909 dòng tương đương 3909 công ty tuy nhiên sau khi lọc và giữ lại các công ty lớn chỉ còn 214 công ty. Tỉ lệ xấp xỉ 1/20. Qua đó có thể thấy .


# 2. Xu hướng tuyển dụng của các công ty lớn hiện nay như thế nào?
Tiêu chí đánh giá công ty lớn: rating, Revenue, Size, Salary (ta đã thực hiện ở trên rồi)

Các vị trí được các công ty này tuyển dụng nhiều nhất.

Lợi ích: 	
+ đối với doanh nghiệp -> nhanh chóng nắm bắt được xu hướng kinh doanh và phát triển của công ty khác
+ đối với các nhà môi giới tuyển dụng -> dễ dàng nắm bắt và nhanh chóng tìm kiếm nguồn nhân sự cho các công ty
+ đối với người tìm việc -> kịp thời nắm bắt xu hướng việc làm để phát triển bản thân và tìm kiếm công việc phù hợp


# Đầu tiên ta cần lọc lại Job Title

In [98]:

df_job_title = df['Job Title'].str.lower()

print(len(df_job_title.value_counts()))

need_replace_1 = [r'.*data analyst.*',r'.*analytics.*',r'.*compliance engineer.*',r'.*modeling engineer.*',r'.*postdoctoral.*',
                r'.*analyst.*', r'.*quality.*', r'.*analytical.*']
need_replace_2 = [r'.*data science.*',r'.*senior scientist.*',r'.*scientist I.*',r'.*engineering.*',r'.*bi scientist.*',
                r'.*innovation.*', r'.*physical scientist.*', r'.*development.*',r'.*bioanalytics.*', r'.*r&d scientist.*',
                r'.*scientist process chemist.*', r'.*rd scientist.*', r'.*data operations.*', r'.*data architect.*',
                r'.*combustion.*', r'.*scientist ii.*', r'.*scientist iv.*', r'.*Scientist iii.*',r'.*data scientist.*',
                 r'.*senior data.*',r'.*senior scientist.*']
need_replace_3 = [r'.*quantitative.*',r'.*research.*',r'.*radar.*']
need_replace_4 = [r'.*ai scientist.*',r'.*learning.*',r'.*robotics scientist.*',r'.*intelligence.*',r'.*informatics scientist.*']

df_job_title = df_job_title.replace(to_replace = r'.*process.*', value = 'Process Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*Ergonomics.*', value = 'Ergonomics Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = [r'.*assistant.*',r'.*experiment support.*'], value = 'Assistant Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*imaging.*', value = 'Imaging Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = [r'.*biopharm.*',r'.*bioinformatics.*'], value = 'Bioinformatics Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*translational.*', value = 'Translational Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*behavioral.*', value = 'Behavioral Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*secure.*', value = 'Secure Data Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*pharmacovgilance.*', value = 'Pharmacovgilance Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*decision.*', value = 'Decision Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*staff.*', value = 'Staff Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*principle.*', value = 'Principle Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*human.*', value = 'Human Factors Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*food.*', value = 'Food Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*technology.*', value = 'Technology Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*project.*', value = 'Project Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*computer.*', value = 'Computer Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*software.*', value = 'Software Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*applications.*', value = 'Applications Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = [r'.*laboratory scientist.*',r'.*lab.*'], value = 'Laboratory Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*data modeler.*', value = 'Data Modeler', regex=True)
df_job_title = df_job_title.replace(to_replace = [r'.*environmental scientist.*',r'.*environmental.*'], value = 'Environmental Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*associate scientist.*', value = 'Associate Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = r'.*applied scientist.*', value = 'Applied Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = [r'.*data consultant.*',r'.*consultant.*'], value = 'Data Consultant', regex=True)
df_job_title = df_job_title.replace(to_replace = [r'.*data engineer.*',r'.*engineering.*'], value = 'Data Engineer', regex=True)
df_job_title = df_job_title.replace(to_replace = need_replace_4, value = 'AI Scientist', regex=True)
df_job_title = df_job_title.replace(to_replace = need_replace_3, value = 'Quantitative Researcher', regex=True)
df_job_title = df_job_title.replace(to_replace = need_replace_1, value = 'Data Analyst', regex=True)
df_job_title = df_job_title.replace(to_replace = need_replace_2, value = 'Data Science', regex=True)

df_job_title.value_counts()

2054


Data Science                                                           1031
Data Analyst                                                            913
Data Engineer                                                           759
AI Scientist                                                            250
Quantitative Researcher                                                 213
                                                                       ... 
drug product scientist                                                    1
upstream scientist - bioreactors                                          1
scientist i metrix                                                        1
scientist - microencapsulation                                            1
patient safety physician or safety scientist - uk, europe or the us       1
Name: Job Title, Length: 182, dtype: int64

2 biểu đồ cột theo (%)